src/llm/visual_language_model/continuous_batching/servable.cpp

```diff
@@ -64,16 +65,31 @@ absl::Status VisualLanguageModelServable::prepareInputs(std::shared_ptr<GenAiSer
     }
     if (executionContext->endpoint == Endpoint::CHAT_COMPLETIONS) {
         ov::genai::ChatHistory& chatHistory = vlmExecutionContext->apiHandler->getChatHistory();
+
+        // Validate chat history for restricted tags
+        for (const auto& historyEntry : chatHistory) {
+            for (const auto& [_, content] : historyEntry) {
+                if (content.find("<ov_genai_image_") != std::string::npos) {
+                    return absl::InvalidArgumentError("Message contains restricted <ov_genai_image> tag");
+                }
+            }
+        }
+
         const ImageHistory& imageHistory = vlmExecutionContext->apiHandler->getImageHistory();
         size_t imageIndex = 0;
+        std::unordered_map<size_t, std::string> imageTags;
         for (const auto& image : imageHistory) {
             const auto& [chatTurnIndex, imageTensor] = image;
             std::string imageTag = "<ov_genai_image_" + std::to_string(imageIndex++) + ">\n";
-            chatHistory[chatTurnIndex]["content"] = imageTag + chatHistory[chatTurnIndex]["content"];
+            imageTags[chatTurnIndex] = imageTags[chatTurnIndex] + imageTag;
             vlmExecutionContext->inputImages.push_back(imageTensor);
         }
+        for (const auto& [chatTurnIndex, imageTagString] : imageTags) {
+            chatHistory[chatTurnIndex]["content"] = imageTagString + chatHistory[chatTurnIndex]["content"];
+        }
         constexpr bool add_generation_prompt = true;  // confirm it should be hardcoded
         vlmExecutionContext->inputText = properties->tokenizer.apply_chat_template(chatHistory, add_generation_prompt);
+        SPDLOG_INFO("Input text:\n{}", vlmExecutionContext->inputText);
```

Comment:
`inputText` is logged below on the trace level

Reply:
removed